N2kMessages.h - added msToMPH to convert meters/second to Miles/hour

forcel0 opened the pull request:
N2kTimer.cpp - added timing support for Pi Pico
NMEA2000_CAN.h - added support for the Raspberry Pi Pico with Adafruit CAN PiCow hat
Seasmart.h - modified include header to search include paths "" vs. <> system-defined

ttlappalainen commented:
- Where is NMEA2000_PICO.h
- How do you think below works?
```
#if defined(ARDUINO_ARCH_ESP32) || defined(ESP_PLATFORM)
  N2kMillis64() and N2kMillis() as inline on header

```

forcel0 commented:
> * Where is NMEA2000_PICO.h
Currently in my repo https://github.com/forcel0/NMEA2K-Pico.  I wasn't sure or organization structure.  It was based of NMEA2000_mcp which is in a separate repository on ttlappalainen. Let me know your thoughts on this.

> * How do you think below works?
> 
> ```
> #if defined(ARDUINO_ARCH_ESP32) || defined(ESP_PLATFORM)
>   N2kMillis64() and N2kMillis() as inline on header
> ```
It doesn't. It worked for me as I have a different #define so I didn't catch it, but it was definitely an error on my end. I've corrected and resubmitted a pull request.

ttlappalainen commented:
- Why there is indention change on N2kTimer.cpp line 52/53?
- Should should provide NMEA2000_Pico and own repository not under some of your project.
- You should provide document at beginnning of NMEA2000_CAN.h, where one can find NMEA2000_Pico as there are for others.
- Are files mcp2515.x and can.h derived from somewhere? If yes and someone has both libraries, there are risk for conflicts, which are often difficult for beginners. In this it may be better to rename them or do as I have done int NMEA2000_Teensyx that everything is in it.
- Check also file structure from NMEA2000_Teensyx for NMEA2000_Pico library. Add ReadMe and version information files.
- You have defined on NMEA2000_CAN.h #define MCP_CAN_RX_BUFFER_SIZE 1 and then on NMEA2000_Pico.h #define MCP_CAN_RX_BUFFER_SIZE 50. So if you first include NMEA2000_CAN.h you will get default buffers size to one.
- Remove referencies like // #if defined(ESP8266) from your library. Since itis specialized to Pico, there is no need to expect it would be compiled for other processors.
- Does this base on some mcp2515 hat for Pico? If yes, I'll prefer to also set default value for interrupt pin. Receiving does not work well without interrupted reading.
- Original NMEA2000_mcp were designed for low memory devices so default buffer sizes are very small. Pico has more memory, so you could set  if (MaxCANReceiveFrames<50 ) MaxCANReceiveFrames = 50;  if (MaxCANSendFrames<50 ) MaxCANSendFrames = 50. In receiving side this gives you up to 25 ms buffer, which you may spent with LCD easily.

ttlappalainen commented:
- There is still unnecessary indention change on N2kTimer.cpp line 52/53?
- There is wring comment "This defines the number of the CAN Interrupt PIN. Default Value 0xff " Does hat has fixed pinout for CS and int?
- I prefer to order files in your NMEA2000_PICO library to src and docs and add version files library.json and library.properties as in NMEA2000_Teensyx. I am not specialist with those, but folders makes it cleaner and version files effects to Arduino ot platformIO information. On updates remember also update those version files.
- Fix also your library copyright year and copywriter.
- Before final merge it is best combine commits to lastone otherwise library will have whole history of thse changes. I think this was what I mean: https://www.freecodecamp.org/news/git-squash-commits/

forcel0 commented:
There is still unnecessary indention change on N2kTimer.cpp line 52/53?
fixed

Does hat has fixed pinout for CS and int? 
yes default CS 20 default INT 21. They can be modified I've fixed INT to use 21 I had disabled while testing with scope

I prefer to order files in your NMEA2000_PICO library to src and docs and add version files library.json and library.properties as in NMEA2000_Teensyx. I am not specialist with those, but folders makes it cleaner and version files effects to Arduino ot platformIO information. On updates remember also update those version files.
Fix also your library copyright year and copywriter.
I can work on this

Before final merge it is best combine commits to lastone otherwise library will have whole history of these changes. I think this was what I mean: https://www.freecodecamp.org/news/git-squash-commits/
I'll read this over

forcel0 commented:
does everything look good at this point or did I miss something.

ttlappalainen commented:
I have not yet tested, but could not find anything visually.

Did you try to combine pulls? Now it shows 6 commits. it should show just one so that in library there is no all unnecessary fixes.

forcel0 commented:
I did try to combine the commits. I the past I done the squash and merge wit the Github UI. I can do it prior to the pull request but I think I'd need to close this pull request squash and then open a new pull request. Is that what you'd like me to do?

ttlappalainen commented:
I have not done it myself, but it should be possible to merge current PR:s to one PR. There was discussion about it with some other PR earlier. Removing current and creating new is the last option.

forcel0 commented:
I have usually squashed when merging the PR using the Github UI as long as the repository is configured to allow.
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests
I believe you can also enforce that all PRs going forward must squash when merged by adjusting the repository's settings

forcel0 commented:
> N2kTimer.cpp - added timing support for Pi Pico NMEA2000_CAN.h - added support for the Raspberry Pi Pico with Adafruit CAN PiCow hat - Seasmart.h - modified include header to search include paths "" vs. <> system-defined

takes care of issue Seasmart.h #include <N2kMsg.h> issue #444
